### PR TITLE
Staging

### PR DESCRIPTION
**Pull request by NikolayRn**
## Proposed changes

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.

If it fixes a bug or resolves a feature request, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Refactor or a modification (non-breaking change)
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

File: apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/unstake.ts [hunk 1]
```diff
@@ -0,0 +1,122 @@
+import { ERC_20_TYPE, NATIVE_TOKEN_TYPE } from '@app/smart-wallets-service/common/constants/tokenTypes'
+import WalletAction from './base'
```
Comment (NikolayRn) [inline apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/unstake.ts:2]: We should use global imports

File: apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/token-receive.ts [hunk 2]
```diff
@@ -0,0 +1,110 @@
+import { NATIVE_FUSE_TOKEN } from '@app/smart-wallets-service/common/constants/fuseTokenInfo'
+import WalletAction from './base'
+import { ERC20Transfer, ERC721Transfer } from '../../interfaces/token-interfaces'
```
Comment (NikolayRn) [inline apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/token-receive.ts:3]: Imports

File: apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/swap.ts [hunk 3]
```diff
@@ -0,0 +1,331 @@
+import { ERC_20_TYPE, NATIVE_TOKEN_TYPE } from '@app/smart-wallets-service/common/constants/tokenTypes'
+import WalletAction from './base'
```
Comment (NikolayRn) [inline apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/swap.ts:2]: Imports

File: apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/stake.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
+import { ERC_20_TYPE, NATIVE_TOKEN_TYPE } from '@app/smart-wallets-service/common/constants/tokenTypes'
+import WalletAction from './base'
```
Comment (NikolayRn) [inline apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/stake.ts:2]: Imports

File: apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/nft-transfer.ts [hunk 5]
```diff
@@ -0,0 +1,41 @@
+import { ERC_721_TYPE } from '@app/smart-wallets-service/common/constants/tokenTypes'
+import WalletAction from './base'
```
Comment (NikolayRn) [inline apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/nft-transfer.ts:2]: Imports

File: apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/native-transfer.ts [hunk 6]
```diff
@@ -0,0 +1,41 @@
+import { NATIVE_TOKEN_TYPE } from '@app/smart-wallets-service/common/constants/tokenTypes'
+import { ERC20Transfer } from '../../interfaces/token-interfaces'
```
Comment (NikolayRn) [inline apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/native-transfer.ts:2]: Imports

File: apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/index.ts [hunk 7]
```diff
@@ -0,0 +1,21 @@
+import WalletAction from './base'
+import ApproveToken from './approve-token'
+import ERC20Transfer from './erc20-transfer'
+import NativeTransfer from './native-transfer'
```
Comment (NikolayRn) [inline apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/index.ts:4]: Imports

File: apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/erc20-transfer.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import { ERC_20_TYPE } from '@app/smart-wallets-service/common/constants/tokenTypes'
+import WalletAction from './base'
```
Comment (NikolayRn) [inline apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/erc20-transfer.ts:2]: Imports

File: apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/approve-token.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
+import { ERC_20_TYPE, ERC_721_TYPE } from '@app/smart-wallets-service/common/constants/tokenTypes'
+import WalletAction from './base'
```
Comment (NikolayRn) [inline apps/charge-smart-wallets-service/src/data-layer/models/wallet-action/approve-token.ts:2]: Imports

File: apps/charge-smart-wallets-service/src/data-layer/data-layer.service.ts [hunk 10]
```diff
@@ -1,41 +1,104 @@
 import { Model, PaginateModel } from 'mongoose'
-import { Inject, Injectable } from '@nestjs/common'
+import { Inject, Injectable, Logger } from '@nestjs/common'
 import { userOpString, walletActionString } from './data-layer.constants'
```
Comment (NikolayRn) [inline apps/charge-smart-wallets-service/src/data-layer/data-layer.service.ts:3]: Imports

File: apps/charge-smart-wallets-service/src/data-layer/data-layer.controller.ts [hunk 11]
```diff
@@ -2,6 +2,7 @@ import { DataLayerService } from '@app/smart-wallets-service/data-layer/data-lay
 import { Body, Controller } from '@nestjs/common'
 import { MessagePattern } from '@nestjs/microservices'
 import { BaseUserOp, UserOp } from '@app/smart-wallets-service/data-layer/interfaces/user-op.interface'
+import { TokenTransferWebhookDto } from '../smart-wallets/dto/token-transfer-webhook.dto'
```
Comment (NikolayRn) [inline apps/charge-smart-wallets-service/src/data-layer/data-layer.controller.ts:5]: Imports

**Review by NikolayRn — COMMENTED** [review]
Except non global imports LGTM

**Review by SametSahin10 — COMMENTED** [review]
There are a lot of changes. The parts that I worked on look good to me. I've briefly skimmed through the other changes. Nothing unusual caught my eye.

About converting relative imports to global imports, I've created this PR: #202.
I've also removed some of the logs which I think are not necessary anymore: #203 

@leonprou @NikolayRn

File: .env.example [hunk 12]
```diff
@@ -70,6 +70,7 @@ SMART_WALLETS_PORT=5007
 SMART_WALLETS_TCP_PORT=8882
 SMART_WALLETS_MONGO_URI=mongodb://mongo:27017/charge-smart-wallets
 SMART_WALLETS_JWT_SECRET=smart_wallets_jwt_secret
+INCOMING_TOKEN_TRANSFERS_WEBHOOK_ID=put_incoming_token_transfers_webhook_id_here
```
Comment (leonprou) [inline .env.example:73]: do we have this set on production environment?

File: apps/charge-apps-service/src/common/config/configuration.ts [hunk 13]
```diff
@@ -1,4 +1,6 @@
 export default () => ({
+  CHARGE_BASE_URL: process.env.CHARGE_BASE_URL,
```
Comment (leonprou) [inline apps/charge-apps-service/src/common/config/configuration.ts:2]: should be cammel cased, please see how we define the other variables.

File: apps/charge-smart-wallets-service/src/common/utils/user-op-parser.ts [hunk 14]
```diff
@@ -1,12 +1,40 @@
 import { Injectable } from '@nestjs/common'
-import { decodeWithCalldata, sigHashFromCalldata, decodeWithEventProps } from '../../../../../libs/common/src/utils/dtools/decodeBySigHash'
-import { EventProps } from '../../../../../libs/common/src/utils/dtools/decodeEvent'
-import { isNil } from 'lodash'
+import { decodeWithCalldata, sigHashFromCalldata } from '../../../../../libs/common/src/utils/dtools/decodeBySigHash'
```
[the remark this answers is not in the document]
Comment (leonprou) [reply apps/charge-smart-wallets-service/src/common/utils/user-op-parser.ts:2]: @NikolayRn

File: apps/charge-smart-wallets-service/src/data-layer/data-layer.controller.ts (hunk 11 above)
Comment (leonprou) [reply apps/charge-smart-wallets-service/src/data-layer/data-layer.controller.ts:5]: @NikolayRn can we change those?

File: apps/charge-smart-wallets-service/src/smart-wallets/services/smart-wallets-aa.service.ts [hunk 15]
```diff
@@ -11,16 +14,27 @@ export class SmartWalletsAAService implements SmartWalletService {
...
+    private configService: ConfigService,
+    private chargeApiService: ChargeApiService
     // private readonly centrifugoAPIService: CentrifugoAPIService,
   ) { }
 
   async auth (smartWalletsAuthDto: SmartWalletsAuthDto) {
     try {
+      this.logger.debug('Recovering the public key...')
       const publicKey = recoverPublicKey(arrayify(hashMessage(arrayify(smartWalletsAuthDto.hash))), smartWalletsAuthDto.signature)
+      this.logger.debug('Recovered the public key.')
```
Comment (leonprou) [inline apps/charge-smart-wallets-service/src/smart-wallets/services/smart-wallets-aa.service.ts:28]: let's remove those

File: docker-compose.full-dev.yml [hunk 16]
```diff
@@ -78,11 +78,13 @@ services:
     depends_on:
       - mongo
       - redis
+      - charge-smart-wallets-service
     environment:
       - REDIS_HOST
       - REDIS_PORT
       - REDIS_PASSWORD
       - REDIS_DB
+      - SMART_WALLETS_JWT_SECRET
```
Comment (leonprou) [inline docker-compose.full-dev.yml:87]: do we have this on cloud?

File: k8s/helm/charts/skandha/templates/secrets.yaml [hunk 17]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
+kind: Secret
+metadata:
```
Comment (leonprou) [inline k8s/helm/charts/skandha/templates/secrets.yaml:3]: Hi @NikolayRn did you checked the secrets as well?

**Review by leonprou — COMMENTED** [review]
Looks good
@NikolayRn please see the requests for changes and comments

File: .env.example (hunk 12 above)
Comment (NikolayRn) [reply .env.example:73]: Yes

File: apps/charge-apps-service/src/common/config/configuration.ts (hunk 13 above)
Comment (NikolayRn) [reply apps/charge-apps-service/src/common/config/configuration.ts:2]: It was done in order to match this getter in charge-api.service: 
Screenshot 2023-11-13 at 11 20 23
For some reason we were unable to get it from .env

**Pull request merged**